### PR TITLE
WebIDL HTMLConstructor support

### LINKED ISSUE #17194: Support HTMLConstructor WebIDL Attribute
spec: https://html.spec.whatwg.org/multipage/#htmlconstructor

part of #9372

### PR DESCRIPTION
**Pull request by cbrewster**
spec: https://html.spec.whatwg.org/multipage/#htmlconstructor

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #17194 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17224)

### CODE REVIEW

**Comment by cbrewster** [comment]
This is getting close to done. I am new to a lot of the jsapi stuff and codegen, so any feedback there would be very helpful!

cc @jdm

File: components/script/dom/bindings/interface.rs [hunk 1]
```diff
@@ -158,6 +169,82 @@ pub unsafe fn create_global_object(
...
+            if constructor.is_null() {
+                return Err(Error::JSFailed);
+            }
+
+            // Callee must be the same constructor object as HTMLElement
+            if constructor.get() != callee.get() {
+                return Err(Error::Type("Active function object is not HTMLElement".to_owned()));
+            }
+        } else {
+            // TODO: Step 5
```
Comment (cbrewster) [inline components/script/dom/bindings/interface.rs]: I'm not quite sure how to do step 5. I think there are two potential options:
1. Get a list of local names from the callee. I am not even sure if this is possible.
2. Get a `ConstructorObject` from the definition's local name and compare it to the `callee`. But we don't generate `GetConstructorObject` for most of the HTML interfaces.

**Comment by nox** [comment]
Changes to `WebIDL.py` need to be versioned as standalone patch files in the repository. But in your case you don't need to do it, because the upstream parser in mozilla-central already supports `[HTMLConstructor]`. Will do a PR to update ours.

**Comment by nox** [comment]
https://github.com/servo/servo/pull/17254

File: components/script/dom/bindings/interface.rs [hunk 2]
```diff
@@ -158,6 +167,26 @@ pub unsafe fn create_global_object(
     JS_FireOnNewGlobalObject(cx, rval.handle());
 }
 
+// https://html.spec.whatwg.org/multipage/dom.html#htmlconstructor
+pub fn create_html_element(window: &Window, call_args: CallArgs) -> Fallible<Root<HTMLElement>> {
```
Comment (jdm) [inline components/script/dom/bindings/interface.rs]: &CallArgs

File: components/script/dom/customelementregistry.rs [hunk 3]
```diff
@@ -228,27 +233,42 @@ impl CustomElementRegistryMethods for CustomElementRegistry {
     }
 }
 
+#[derive(HeapSizeOf, JSTraceable, PartialEq, Clone)]
+pub enum ConstructionEntry {
+    Element(Root<HTMLElement>),
```
Comment (jdm) [inline components/script/dom/customelementregistry.rs]: This should be `JS<HTMLElement>`, and the type should be marked `#[must_root]`.

File: components/script/dom/bindings/interface.rs [hunk 4]
```diff
@@ -212,14 +213,34 @@ pub unsafe fn html_constructor(window: &Window, call_args: CallArgs) -> Fallible
...
+
+            // TODO: Steps 8.3-8.4
+
+            // Step 8.5
+            Root::downcast(element).ok_or(Error::InvalidState)
+        },
+        Some(entry) => {
+            let result = match *entry {
+                ConstructionEntry::AlreadyConstructed => return Err(Error::InvalidState),
+                ConstructionEntry::Element(ref mut element) => {
```
Comment (jdm) [inline components/script/dom/bindings/interface.rs]: No need for mut.

**Review by jdm — CHANGES_REQUESTED** [review]
Good start!

**Comment by jdm** [comment]
```
failures:

    size_of::size_div

    size_of::size_element

    size_of::size_htmlelement

    size_of::size_span
```

File: components/script/dom/bindings/interface.rs (hunk 1 above)
Comment (cbrewster) [inline components/script/dom/bindings/interface.rs]: We only generate GetConstructorObject for interfaces that are a namespace, callback, or have descendants.
[Configuration.py#L397-L401](https://github.com/servo/servo/blob/85585d93345be93588545c2c7c6766421c12602f/components/script/dom/bindings/codegen/Configuration.py#L397-L401)

File: components/script/dom/customelementregistry.rs [hunk 5]
```diff
@@ -37,7 +39,8 @@ pub struct CustomElementRegistry {
 
     element_definition_is_running: Cell<bool>,
 
-    definitions: DOMRefCell<HashMap<DOMString, CustomElementDefinition>>,
+    #[ignore_heap_size_of = "Rc"]
+    definitions: DOMRefCell<HashMap<DOMString, Rc<CustomElementDefinition>>>,
```
Comment (cbrewster) [inline components/script/dom/customelementregistry.rs]: The spec<sup>1</sup> says each `Element` has an associated definition. It seems like `Rc` would be the best way to go about that.

1: https://dom.spec.whatwg.org/#concept-element-custom-element-definition

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 6]
```diff
@@ -5291,31 +5291,24 @@ def definition_body(self):
         if self.constructor.isHTMLConstructor():
```
Comment (cbrewster) [inline components/script/dom/bindings/codegen/CodegenRust.py]: Will do! Didn't take the time to cleanup the commits yet

File: components/script/dom/bindings/interface.rs (hunk 1 above)
Comment (jdm) [inline components/script/dom/bindings/interface.rs]: Gecko is much more liberal; we can follow their lead: https://dxr.mozilla.org/mozilla-central/source/dom/bindings/Codegen.py#12826-12828

Comment (nox) [inline components/script/dom/bindings/interface.rs]: Yeah, not generating `GetConstructorObject` for leaf interfaces was just me trimming a bit of unneeded codegen, but if it's not unneeded anymore, just make it generate the function again.

File: components/script/dom/bindings/utils.rs [hunk 7]
```diff
@@ -534,3 +601,146 @@ impl AsCCharPtrPtr for [u8] {
         self as *const [u8] as *const c_char
     }
 }
+
+/// Returns the constructor object for the element associated with the given local name.
+/// This list should only include elements marked with the [HTMLConstructor] extended attribute.
+pub fn get_constructor_object_from_local_name(name: LocalName,
```
Comment (jdm) [inline components/script/dom/bindings/utils.rs]: Let's move this to dom/bindings/interfaces.rs.

**Review by jdm — APPROVED** [review]
r=me once the function is moved.

**Pull request merged**